zallet/src/components/json_rpc/balance.rs

```diff
@@ -0,0 +1,293 @@
+use zcash_protocol::{consensus::BlockHeight, value::Zatoshis};
+
+use crate::components::database::DbConnection;
+
+/// Coinbase transaction outputs can only be spent after this number of new blocks
+/// (network rule).
+const COINBASE_MATURITY: u32 = 100;
+
+enum IsMine {
+    Spendable,
+    WatchOnly,
+    Either,
+}
+
+/// Returns `true` if this output is owned by some account in the wallet and can be spent.
+pub(super) fn is_mine_spendable(
+    wallet: &DbConnection,
+    tx_out: &TxOut,
+) -> Result<bool, SqliteClientError> {
+    is_mine(wallet, tx_out, IsMine::Spendable)
+}
+
+/// Returns `true` if this output is owned by some account in the wallet, but cannot be
+/// spent (e.g. because we don't have the spending key, or do not know how to spend it).
+#[allow(dead_code)]
+pub(super) fn is_mine_watchonly(
+    wallet: &DbConnection,
+    tx_out: &TxOut,
+) -> Result<bool, SqliteClientError> {
+    is_mine(wallet, tx_out, IsMine::WatchOnly)
+}
+
+/// Returns `true` if this output is owned by some account in the wallet.
+pub(super) fn is_mine_spendable_or_watchonly(
+    wallet: &DbConnection,
+    tx_out: &TxOut,
+) -> Result<bool, SqliteClientError> {
+    is_mine(wallet, tx_out, IsMine::Either)
+}
+
+/// Logically equivalent to [`IsMine(CTxDestination)`] in `zcashd`.
+///
+/// [`IsMine(CTxDestination)`]: https://github.com/zcash/zcash/blob/2352fbc1ed650ac4369006bea11f7f20ee046b84/src/script/ismine.cpp#L121
+fn is_mine(
+    wallet: &DbConnection,
+    tx_out: &TxOut,
+    include: IsMine,
+) -> Result<bool, SqliteClientError> {
+    match tx_out.recipient_address() {
+        Some(address) => wallet.with_raw(|conn| {
+            let mut stmt_addr_mine = conn.prepare(
+                "SELECT EXISTS(
+                    SELECT 1
+                    FROM addresses
+                    JOIN accounts ON account_id = accounts.id
+                    WHERE address = :address
+                    AND (
+                        :allow_either = 1
+                        OR has_spend_key = :has_spend_key
+                    )
+                )",
+            )?;
+
+            Ok(stmt_addr_mine.query_row(
+                named_params! {
+                    ":address": address.encode(wallet.params()),
+                    ":allow_either": matches!(include, IsMine::Either),
+                    ":has_spend_key": matches!(include, IsMine::Spendable),
+                },
+                |row| row.get(0),
+            )?)
+        }),
+        // TODO: Use `zcash_script` to discover other ways the output might belong to
+        // the wallet (like `IsMine(CScript)` does in `zcashd`).
+        None => Ok(false),
+    }
+}
+
+/// Equivalent to [`CTransaction::GetValueOut`] in `zcashd`.
+///
+/// [`CTransaction::GetValueOut`]: https://github.com/zcash/zcash/blob/2352fbc1ed650ac4369006bea11f7f20ee046b84/src/primitives/transaction.cpp#L214
+pub(super) fn wtx_get_value_out(tx: &Transaction) -> Option<Zatoshis> {
```

Comment:
The semantics of this whole method are very weird; how would someone reasonably use this value?

Reply:
Notice the TODO at https://github.com/zcash/wallet/pull/36/files#diff-aa4c5eb784ec34fa821ac213851b030ecd0d609b8d5b9f94252db39e476b68f6R191 , just before the only call to this method. I think the best way to proceed is to remove this method and fix that TODO in the way it suggests.